### PR TITLE
Angular Crud Book App - Yan Yin Choy

### PR DESCRIPTION
**Pull request by ychoy**
Challenges: Going through all the steps to get a route set up e.g. making sure it's connected on app.js, index.html, etc. 
Complete: 4.75/5  -- Got cancel button to work but it doesn't redirect to homepage...Need to write a function that works and include on BooksShowController.js
Thoughts: I need to review Angular docs to understand other services that I can call on and integrate with bootstrap.  And save the styling for last...

### CODE REVIEW

File: public/scripts/controllers/BooksIndexController.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  function BooksIndexController ($http) {
+    var vm = this;
+
+      $http({
+        method: 'GET',
+        url: 'https://super-crud.herokuapp.com/books'
+      }).then(function onBooksIndexSuccess(response) {
+        vm.books = response.data.books;
+      }, function onError(response) {
```
Comment (thelmaboamah) [inline public/scripts/controllers/BooksIndexController.js:15]: You're not doing anything with the error response here.

**Review by thelmaboamah — APPROVED** [review]
Looks good. Error callbacks are the only thing I noted.

File: index.html [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html>
+<html lang="en" ng-app="booksApp">
+<head>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1">
+
+  <title>Books App</title>
+
+ 	 <!-- STYLESHEETS -->
```
Comment (mnfmnfm) [inline index.html:9]: weird indentation

File: index.html [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+   <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.5.6/angular.min.js"></script>
+   <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.5.6/angular-route.js"></script>
+   <!-- APPLICATION SCRIPTS -->
+   <script type="text/javascript" src="../public/scripts/app.js"></script>
+   <script type="text/javascript" src="../public/scripts/controllers/BooksIndexController.js"></script>
+   <script type="text/javascript" src="../public/scripts/controllers/BooksShowController.js"></script>
+
+</head>
+<body>
+  <div class="jumbotron">
```
Comment (mnfmnfm) [inline index.html:23]: why do you need both jumbotron and container here?

File: public/scripts/app.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+angular
+  .module('booksApp', ['ngRoute']) // ngRoute allows us to configure client-side routes in our app.
+  .config(config); // for the app, use this dependency and configuration
+
+/* ROUTES */
+
+  config.$inject = ['$routeProvider', '$locationProvider'];
```
Comment (mnfmnfm) [inline public/scripts/app.js:7]: This shouldn't be indented, and it throws off you indentation for the rest of the file.

File: public/scripts/app.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        controllerAs: 'booksIndexCtrl',
+        controller: 'BooksIndexController'
+      })
+      .when('/books/:id', {
+        templateUrl: '/templates/books-show.html',
+        controllerAs: 'booksShowCtrl',
+        controller: 'BooksShowController'
+      })
+      .otherwise({
+      redirectTo: '/'
```
Comment (mnfmnfm) [inline public/scripts/app.js:22]: this should be more indented to line up w/ others

File: public/scripts/controllers/BooksIndexController.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      });
+
+      vm.createBook = function () {
+        $http({
+          method: 'POST',
+          url: 'https://super-crud.herokuapp.com/books',
+          data: vm.newBook,
+        }).then(function successCallback(response) {
+          vm.books.push(response.data);
+        }, function errorCallback(response) {
```
Comment (mnfmnfm) [inline public/scripts/controllers/BooksIndexController.js:25]: this should be indented one more level

File: public/scripts/controllers/BooksShowController.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+BooksShowController.$inject=['$http','$routeParams', '$location'];
+
+function BooksShowController($http, $routeParams, $location) {
+  var vm = this;
+  var bookId = $routeParams.id
+  vm.books = [];
+  $http({
+    method: 'GET',
+    url: 'https://super-crud.herokuapp.com/books/'+ bookId
```
Comment (mnfmnfm) [inline public/scripts/controllers/BooksShowController.js:13]: should use a template string instead of concatenation

File: public/scripts/controllers/BooksShowController.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
+angular
+  .module('booksApp')
+  .controller('BooksShowController', BooksShowController)
+
+BooksShowController.$inject=['$http','$routeParams', '$location'];
+
+function BooksShowController($http, $routeParams, $location) {
+  var vm = this;
+  var bookId = $routeParams.id
```
Comment (mnfmnfm) [inline public/scripts/controllers/BooksShowController.js:9]: missing semicolon

File: public/scripts/controllers/BooksShowController.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    url: 'https://super-crud.herokuapp.com/books/'+ bookId
+  }).then(function onBookShowSuccess(response) {
+    vm.book = response.data;
+  }, function onError(response){
+ });
+
+ vm.updateBook = function(bookToUpdate) {
+     $http({
+       method: 'PUT',
+       url: 'https://super-crud.herokuapp.com/books/'+bookToUpdate._id ,
```
Comment (mnfmnfm) [inline public/scripts/controllers/BooksShowController.js:22]: same comment about template vs. concat

File: public/scripts/controllers/BooksShowController.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+       data: {
```
Comment (mnfmnfm) [inline public/scripts/controllers/BooksShowController.js:23]: This isn't wrong, but you could just send bookToUpdate as your data object.

File: public/scripts/controllers/BooksShowController.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+       vm.book = response.data; // save the updated book to the DB
+       $location.path('/'); // redirect to homepage after book is saved
+     }, function onError(response){
+     });
+  }
+
+  vm.deleteBook = function (bookToDelete) {
+    $http({
+      method: 'DELETE',
+      url: 'https://super-crud.herokuapp.com/books/'+ bookToDelete._id,
```
Comment (mnfmnfm) [inline public/scripts/controllers/BooksShowController.js:39]: same

File: templates/books-index.html [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+                <button class="btn btn-primary">Submit</button>
+              </div>
+            </div>
+          </fieldset>
+        </form>
+      </div>
+  </div>
+
+
+  <input type="text" class="form-control" ng-model="searchText" placeholder="Search Books">
```
Comment (mnfmnfm) [inline templates/books-index.html:57]: Using placeholder instead of labels is bad. See https://www.nngroup.com/articles/form-design-placeholders/

File: templates/books-index.html [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+  <div ng-repeat="book in booksIndexCtrl.books | orderBy: 'title' | filter: searchText" class="row">
+      <br>
```
Comment (mnfmnfm) [inline templates/books-index.html:59]: weirdly indented

File: templates/books-show.html [hunk 14]
```diff
@@ -0,0 +1,49 @@
+<div class="container">
+ <div class="row">
+   <div class="col-md-10 col-md-offset-1">
+     <div class="row">
+       <div class="col-md-10 col-md-offset-1">
```
Comment (mnfmnfm) [inline templates/books-show.html:5]: why are there so many divs in divs in divs that you never use???? this reeks of copy/pasting and never looking back at your html.

**Review by mnfmnfm — COMMENTED** [review]
Has some weirdness in the amount of divs in your HTML, but otherwise good.
